### PR TITLE
`lib.fileset.difference`: init

### PR DESCRIPTION
**Pull request by infinisil**
## Description of changes

This is another split off from the [WIP file set combinators PR](https://github.com/NixOS/nixpkgs/pull/222981). This PR adds one function:
- `lib.fileset.difference`: Subtracts the second file set from the first.
  ```nix
  # Recursively includes all files in ./. except ones under ./a/b
  lib.fileset.difference ./. ./a/b
  ```

This work is sponsored by [Antithesis](https://antithesis.com/) :sparkles:

## Things done

- [x] Documentation
- [x] Tests

### CODE REVIEW

**Review by fricklerhandwerk — APPROVED** [review]

File: lib/fileset/default.nix
Comment (roberth) [inline lib/fileset/default.nix:337]:
```suggestion
        # A set of Nix-related files
```

`nix/` may contain non-`.nix` files, which could be considered non-Nix.

File: lib/fileset/internal.nix
Comment (roberth) [inline lib/fileset/internal.nix:658]:
```suggestion
      # The filesetTree's need to have the same base path in order to be able to compute the difference between them.
      # Technically an unrelated root in the second, negative argument could be considered not to match any files in
      # the first argument, but users would be clueless as to why their filter doesn't work.
```

File: lib/fileset/internal.nix
Comment (roberth) [inline lib/fileset/internal.nix:662]: Not clear to me why.
What is `tree2` for?

File: lib/fileset/tests.sh
Comment (roberth) [inline lib/fileset/tests.sh:728]: How about removing the parent? e.g.

```nix
difference ./a ./.
```

Comment (roberth) [reply lib/fileset/tests.sh:728]: Maybe the test could be more systematic, by framing it as a matrix of possible arguments. Perhaps not an actual matrix in terms of code, but laid out in a way that you can see that all combinations are tested.
That way you would naturally test `difference ./a ./.`.

File: lib/fileset/internal.nix
Comment (infinisil) [reply lib/fileset/internal.nix:658]: Hmm not 100% sure what you mean here. If the root of the second argument is unrelated (e.g. `difference /foo /bar`), then no files inside `/foo` match `/bar`, meaning nothing gets subtracted, so the result is `/foo`. And that should be expected.

File: lib/fileset/internal.nix
Comment (infinisil) [reply lib/fileset/internal.nix:662]: Same as the tree of the negative file set, but with the same base path as the positive file set, such that they are comparable and can be operated on.

E.g. if you have `difference /foo /foo/bar`, the common path would be `/foo`, meaning the first branch applies.

It then takes the negative file set, `(base: /foo/bar, tree: "directory")` in this case, and shortens the base to `/foo` instead, such that you get `(base: /foo, tree: { bar = "directory"; })`. `tree2` here only contains the tree part though, so `{ bar = "directory"; }`.

Now we have two trees, `"directory"` from the positive file set, and `{ bar = "directory"; }` from the negative file set, that have the same base `/foo`, so they can be operated on.

I'll put something like this in the code docs, and I'll improve the variable name, `negativeTreeWithPositiveBasePath` or so (maybe shorter :P)

File: lib/fileset/internal.nix
Comment (roberth) [reply lib/fileset/internal.nix:658]: I was thinking of entirely unrelated trees, such as when one is an absolute path, but the other is a `fetchGit` lazy tree.

Rejecting `difference /foo /bar` seems excessive, I agree.

Comment (infinisil) [reply lib/fileset/internal.nix]: Ah yeah, it will error for different filesystem roots, there's a comment on this in the comments for `_difference`:

```
# The filesets must already be coerced and validated to be in the same filesystem root.
```

The public `difference` function passes its arguments through `_coerceMany`, which checks against this, before calling the internal one.

File: lib/fileset/tests.sh
Comment (infinisil) [reply lib/fileset/tests.sh:778]: Did that!

**Comment by infinisil** [comment]
@roberth Should be good now!

File: lib/fileset/default.nix
Comment (roberth) [inline lib/fileset/default.nix:417]:
```suggestion
        {
          context = "first argument (positive set)";
          value = positive;
        }
        {
          context = "second argument (negative set)";
          value = negative;
        }
```

Not 100% sure, but might be nice.

File: lib/fileset/internal.nix
Comment (roberth) [inline lib/fileset/internal.nix:667]:
```suggestion
      negativeTreeWithPositiveBase =
```

**Review by roberth — APPROVED** [review]

**Comment by infinisil** [comment]
Done now, unless you do it before me, I'll merge once CI passes :)

**Pull request merged**
